Enhanced nav mobile optimizations

miguelaenlle opened the pull request.

miguelaenlle commented on apps/prairielearn/assets/scripts/pageLayoutClient.ts:
```diff
@@ -65,4 +71,42 @@ onDocumentReady(async () => {
     });
   });
+
+  sideNavMobileButton.addEventListener('click', async () => {
```
There's lots of repeat code between the sideNavMobileButton and the regular side nav button. Is there a better way to do this? (Maybe sharing the ID?)

miguelaenlle commented on apps/prairielearn/assets/scripts/pageLayoutClient.ts:
```diff
@@ -65,4 +71,42 @@ onDocumentReady(async () => {
+      courseNavContent.classList.remove('collapse');
+      courseNavContent.classList.add('collapsing');
+      // Collapse the course nav
+      // courseNavContent.classList.remove('show');
```
Remove the commented out code

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,185 @@ body:has(div.app-container) {
+}
+
+/* Side nav collapsed styles do not appear on small/mobile screen sizes */
+@media (min-width: 768px) {
```
Clarify why we don't show the collapsed navbar on smaller screen sizes: because it only pops out from the side

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,185 @@ body:has(div.app-container) {
+    padding-top: 1rem;
+    display: block;
+    transition: transform 0.3s ease-in-out;
+    transform: translateX(-100%); /* Start off-screen */
```
off the screen

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,185 @@ body:has(div.app-container) {
+    z-index: 999;
+  }
+
+  #course-nav {
```
- Explain why we apply these styles specifically to course-nav
- Apply these styles only if side nav is enabled

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,185 @@ body:has(div.app-container) {
+    transform: translateX(100%);
+  }
+
+  /* Prevent main content from being visible when either side menu is open */ 
```
Clean this comment up

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,185 @@ body:has(div.app-container) {
+  }
+
+  /* Prevent main content from being visible when either side menu is open */ 
+  #app-container:has(#course-nav.show) .app-main,
```
Ask about :has -- it's not available on older browsers: https://developer.mozilla.org/en-US/docs/Web/CSS/:has#browser_compatibility

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,185 @@ body:has(div.app-container) {
+    overflow-y: auto;
+  }
+  
+  body:has(#app-container:not(.mobile-collapsed)) .tooltip {
```
Comment why we do this

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -70,22 +75,39 @@ export function Navbar({
           data-bs-toggle="collapse"
-          data-bs-target=".navbar-collapse"
+          data-bs-target="#course-nav"
+          data-bs-delay="0"
```
This does nothing

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -103,7 +125,7 @@ export function Navbar({
                   id="navbar-load-from-disk"
                   class="btn btn-success btn-sm"
-                  href="${urlPrefix}/loadFromDisk"
+                  href="${resLocals.urlPrefix}/loadFromDisk"
```
Rationale for this change? This does exactly what the previous code did...

miguelaenlle commented on apps/prairielearn/src/components/PageLayout.html.ts:
```diff
@@ -113,6 +113,7 @@ export function PageLayout({
               'app-container',
               sideNavEnabled && 'side-nav-enabled',
               !sideNavExpanded && 'collapsed',
+              'mobile-collapsed',
```
Comment what mobile-collapsed is: separate collapse state for mobile. The navbar collapse state is not persisted on mobile devices.

miguelaenlle reviewed (commented):
Initial self-review.

nwalters512 commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -10,6 +10,7 @@ body:has(div.app-container) {
+  position: relative;
 }
 
 /* Styles when the side nav is enabled and expanded */
```
I can't comment on the exact line, but: with this PR, on mobile devices, there seems to be an issue with horizontal scrolling, where it tries to scroll unnecessarily. In this screenshot from Chrome, notice the horizontal scrollbar and the person icon being cut off:

<img width="449" alt="Screenshot 2025-05-19 at 16 29 04" src="https://github.com/user-attachments/assets/bf505727-0838-4565-8b64-39488e9ff61f" />

For me, this was resolvable by changing `width: 100vw` to `width: 100%` a few lines below this.

miguelaenlle replied:
Yup, that fixed it on my end as well:

![image](https://github.com/user-attachments/assets/93e1142f-9ef8-4239-99cb-8b278a7c921a)

nwalters512 commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,164 @@ body:has(div.app-container) {
+    overflow-x: hidden;
+  }
+
+  /* If the side nav is collapsed on a wider viewports,
+  *  tooltips appear above the side nav link icons when hovered. 
+  *  
+  *  If the user collapses the side nav on a wide viewport, then 
+  *  resizes their window to a narrow viewport (less than 768px), 
+  *  the tooltips should be hidden, since side nav link text
+  *  is displayed on narrow viewports. */
```
Formatting/description nits:

```suggestion
  /**
   * On wider viewports, when the side nav is collapsed,
   * tooltips appear next to the side nav link icons when hovered. 
   *
   * This rule hides those tooltips on narrow viewports.
   */
```

nwalters512 commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -198,6 +127,164 @@ body:has(div.app-container) {
+  *  resizes their window to a narrow viewport (less than 768px), 
+  *  the tooltips should be hidden, since side nav link text
+  *  is displayed on narrow viewports. */
+  body:has(.app-container.side-nav-enabled:not(.mobile-collapsed)) .tooltip {
```
Is the `:not(.mobile-collapsed)` bit needed here? I'd think we'd unconditionally want to hide these tooltips on narrow displays.

Very importantly: I'm pretty sure this hides _every_ tooltip under `body`, not just the ones in the sidebar that you care about, right?

miguelaenlle replied:
This hides every tooltip under `body`. Because of this, I used `:not(.mobile-collapsed)` so that other tooltips than the side nav tooltips will be shown if the side nav is collapsed. Let me know if that's okay, or if I should remove `:not(.mobile-collapsed)`.

nwalters512 replied:
Ah, OK, I think I see better what's happening now. So, this does still impact every tooltip on the page, but that doesn't matter because the rest of the page isn't visible when the sidebar isn't collapsed (that is, is expanded) on mobile?

miguelaenlle replied:
Yes. I also set it up this way since tooltips anywhere, including for the sidebar, are rendered as immediate children of the `body`: 

![Screenshot 2025-05-21 at 3 31 06 PM](https://github.com/user-attachments/assets/233d9e4a-eec1-454d-8964-75c0092bd6fa)

nwalters512 replied:
That makes sense, thanks for clarifying!

nwalters512 reviewed (commented):
- There's a `me-2` on `li.nav-item.dropdown` that I think can be removed, it makes the dropdown appear with consistent left/right margins on mobile when it's expanded.
- There's a bit of jank when the window is resized, with the sidebar jumping on/off the screen. It's especially noticable when transitioning from a larger viewport to a smaller one, when the sidebar momentarily fills the entire width of the window before animating away. Admittedly this probably won't impact most people, but it could impact folks who are trying to tile multiple PL windows next to each other, e.g. for question editing. IMO it'd be preferable to only animate the transition when someone explicitly clicks the expand/collapse button.
- Relating to the above: we should respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion). I'm happy if you want to leave this for another PR, but we should do it, this is important for accessibility.

nwalters512 commented:
The flakey test here should be resolved by https://github.com/PrairieLearn/PrairieLearn/pull/12017.

nwalters512 commented on apps/prairielearn/assets/scripts/pageLayoutClient.ts:
```diff
@@ -65,4 +84,56 @@ onDocumentReady(async () => {
+      // Respond to the end of the animation
+      appContainerDiv.addEventListener('transitionend', () => {
+        setTimeout(() => {
+          appContainerDiv.classList.toggle('mobile-collapsed-animating');
+          appContainerDiv.classList.toggle('mobile-collapsed');
```
Nit: here and in the `setTimeout` below, would it be at least as correct but maybe more explicit if we used `classList.remove` here? That is, would we ever want a transition from "no class" to "has class", or do we just care about the removal behavior of `classList.toggle`?

miguelaenlle replied:
I agree with using `remove`/`add` rather than `toggle`. I switched back to `remove`/`add`; I was testing a `toggle`-based implementation to see if it would resolve an issue with the side/user menus collapsing/expanding incorrectly, but I resolved it separately.

nwalters512 reviewed (commented):
Another UX note: at least for me, the expand/collapse of the nav sidebar is basically instant (or at least, the animation starts as soon as I click the button). However, for the user menu, there seems to be a small delay before the animation start. Is that reproducible for you too? Do you have any idea what's causing that, maybe some misconfiguration with a CSS animation?

miguelaenlle commented on apps/prairielearn/assets/scripts/pageLayoutClient.ts:
```diff
@@ -65,4 +84,69 @@ onDocumentReady(async () => {
   });
+
+  sideNavMobileButton.addEventListener('click', async () => {
+    // Animate the side nav
```
Animate -> Transition so the two are not confused?

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -228,13 +232,21 @@ body:has(div.app-container) {
     transform: translateX(-100%); /* Start off-screen to the left */
-    will-change: transform; /* Performance optimization */
     min-height: 100vh;
+    transition: none;
```
Necessary?

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -244,29 +256,31 @@ body:has(div.app-container) {
   /* Prevent content under the course/side nav from being visible when
   either nav is open. */
-  .app-container.side-nav-enabled:has(#course-nav.show) .app-main,
+  .app-container.side-nav-enabled:has(#course-side-nav.show) .app-main,
```
`.show` won't do anything -- `course-side-nav` uses `.mobile-collapsed`

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -289,35 +303,10 @@ body:has(div.app-container) {
   }
 }
 
-#assessment-switcher-container .assessment-row {
```
Regression: lost some pageLayout code here...

miguelaenlle commented on apps/prairielearn/src/components/Navbar.html.ts:
```diff
@@ -38,6 +38,31 @@ export function Navbar({
   navbarType ??= resLocals.navbarType;
 
+  const navbarContent = html`
+    <ul class="nav navbar-nav me-auto" id="main-nav">
```
Should `nav navbar-nav` be in here? I don't think so...

miguelaenlle reviewed (commented):
Comments from self-review

miguelaenlle commented on apps/prairielearn/assets/scripts/pageLayoutClient.ts:
```diff
@@ -65,4 +84,69 @@ onDocumentReady(async () => {
+
+    const courseNavExpanded = !courseNavDiv.classList.contains('mobile-collapsed');
+
+    // Animate the course side nav
```
Replace "course side nav" with "course nav" (same below)

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -233,4 +237,106 @@ body:has(div.app-container) {
+    transition: transform 300ms;
+  }
+
+  .app-container.side-nav-enabled #course-nav {
```
Remove this style; it is overridden by the other one. You might be able to remove the !Important after removing this.

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -233,4 +237,106 @@ body:has(div.app-container) {
+    background-color: rgba(var(--bs-dark-rgb));
+    padding: 1rem 0.5rem;
+    will-change: transform; /* Performance optimization */
+    transform: translateX(100%); /* Start off-screen to the right */
```
Does this do anything? By default mobile-collapsed is already added

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -233,4 +237,106 @@ body:has(div.app-container) {
+    z-index: 999;
+    background-color: rgba(var(--bs-dark-rgb));
+    padding: 1rem 0.5rem;
+    will-change: transform; /* Performance optimization */
```
Should  this be moved under the `.animate` styles for `course-nav`?

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -233,4 +237,106 @@ body:has(div.app-container) {
+    overflow-y: hidden;
+  }
+
+  .app-container.side-nav-enabled:has(#course-nav:not(.mobile-collapsed)),
```
Add a comment above this

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -233,4 +237,106 @@ body:has(div.app-container) {
+    display: none;
+  }
+
+  .viewtype-dropdown-item {
```
Does this apply only when side nav is enabled? If so, we should specify that; if not, this should be moved to Bootstrap

miguelaenlle reviewed (commented):
Self-review comments.

miguelaenlle commented:
> * There's a `me-2` on `li.nav-item.dropdown` that I think can be removed, it makes the dropdown appear with consistent left/right margins on mobile when it's expanded.

I removed the `me-2`, and that fixed this issue:

![image](https://github.com/user-attachments/assets/b8722ec9-c116-4ca8-9adb-57ad4cc1c322)

> * There's a bit of jank when the window is resized, with the sidebar jumping on/off the screen. It's especially noticable when transitioning from a larger viewport to a smaller one, when the sidebar momentarily fills the entire width of the window before animating away. Admittedly this probably won't impact most people, but it could impact folks who are trying to tile multiple PL windows next to each other, e.g. for question editing. IMO it'd be preferable to only animate the transition when someone explicitly clicks the expand/collapse button.

This issue was also happening on my end. I fixed this by adding an `animate` class which enables the CSS transition. When the user clicks the side or course nav toggle button, `animate` is added to `app-container`. 

The side and course nav then have transitions enabled when the `animate` class is added:

https://github.com/PrairieLearn/PrairieLearn/blob/494f720403d810b54c3ae3079e27b513d1c13754/apps/prairielearn/assets/stylesheets/pageLayout.css#L270-L273

https://github.com/PrairieLearn/PrairieLearn/blob/494f720403d810b54c3ae3079e27b513d1c13754/apps/prairielearn/assets/stylesheets/pageLayout.css#L288-L291

When the transition ends, the `animate` class is removed:

https://github.com/PrairieLearn/PrairieLearn/blob/494f720403d810b54c3ae3079e27b513d1c13754/apps/prairielearn/assets/scripts/pageLayoutClient.ts#L109-L116

https://github.com/PrairieLearn/PrairieLearn/blob/494f720403d810b54c3ae3079e27b513d1c13754/apps/prairielearn/assets/scripts/pageLayoutClient.ts#L138-L145

The issue is now mostly resolved:

https://github.com/user-attachments/assets/c9032e6d-390c-4a47-ac92-4e59eb56f92f

There's a small error wherein the side nav animates if it's collapsed and resized:

https://github.com/user-attachments/assets/1180322c-6028-40a2-a840-85324e3e3b63

@nwalters512 , do you think this is okay to leave as-is, or would you recommend finding a way to prevent animation in this case too?

> * Relating to the above: we should respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion). I'm happy if you want to leave this for another PR, but we should do it, this is important for accessibility.

I added some CSS to respect `prefers-reduced-motion`:

https://github.com/user-attachments/assets/583b7e2f-1052-4d9b-8823-8972798335ee

miguelaenlle commented:
> Another UX note: at least for me, the expand/collapse of the nav sidebar is basically instant (or at least, the animation starts as soon as I click the button). However, for the user menu, there seems to be a small delay before the animation start. Is that reproducible for you too? Do you have any idea what's causing that, maybe some misconfiguration with a CSS animation?

Yup, I had that issue as well. I addressed this by having client-side JS control a `mobile-collapsed` class, which determines if the `user-menu` is collapsed/expanded:

https://github.com/PrairieLearn/PrairieLearn/blob/494f720403d810b54c3ae3079e27b513d1c13754/apps/prairielearn/assets/scripts/pageLayoutClient.ts#L125-L152

In the previous implementation, the CSS responded to the collapsed state controlled by Bootstrap toggle, which has a slight delay between the button click and the collapsed class being added. That functionality is still preserved if the side nav is not enabled.

Both the side and user menu toggle buttons are now instantaneous:

https://github.com/user-attachments/assets/8688dad2-bd84-48ca-a459-93ce1f24e0b4

miguelaenlle commented:
> @miguelaenlle I resolved the merge conflicts in [dfab2ef](https://github.com/PrairieLearn/PrairieLearn/commit/dfab2ef1304bd514c61b84d0bb91cfb0b2f09d49). Can you double-check that I did so correctly?

Thank you, @nwalters512 . I'll take a look 👍

miguelaenlle commented:
Also, the expand/collapse side nav button had a padding adjustment in #12134 that aligned the border below the icon with the secondary navigation tabs' bottom border. This adjustment is no longer present:

![Screenshot 2025-06-10 at 4 06 34 PM](https://github.com/user-attachments/assets/d34639fd-9d14-40e1-9554-90ba5860bc36)

I'll work on resolving this issue 👍

miguelaenlle commented on apps/prairielearn/assets/stylesheets/pageLayout.css:
```diff
@@ -10,35 +10,22 @@ body:has(div.app-container) {
   height: 100%;
-  width: 100vw;
+  width: 100%;
+  overflow-x: hidden;
```
I opened an issue -- https://github.com/PrairieLearn/PrairieLearn/issues/12163 -- and a PR that resolves it -- #12165   👍

miguelaenlle commented:
> Also, the expand/collapse side nav button had a padding adjustment in https://github.com/PrairieLearn/PrairieLearn/pull/12134 that aligned the border below the icon with the secondary navigation tabs' bottom border. This adjustment is no longer present:

This issue is now resolved:

![Screenshot 2025-06-11 at 9 40 22 AM](https://github.com/user-attachments/assets/801e9ab7-bb32-4b28-9bee-3836b0834e26)

miguelaenlle commented:
In my testing, an issue came up wherein if the user opened the user dropdown menu on desktop, then resized to mobile with both the left and right sidebars collapsed, opening the left sidebar would leave them stuck on the right sidebar:

![Screenshot 2025-06-11 at 11 50 20 AM](https://github.com/user-attachments/assets/3eac450e-c634-4aa4-8d23-7355547c3091)

To address this, I adjusted the behavior of the right sidebar: when the user opens the user dropdown menu on desktop, the right sidebar opens on mobile. If they resize to mobile, the right sidebar will be open, and the glitch no longer occurs.

nwalters512 reviewed (approved):
I played with this locally, it's working nicely! Let's ship this.
